### ISSUE TITLE
CKEditor - Creating a node with an inline image does not activate CKEditor 5

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#6591]
## Description of the bug

I first noticed this problem on forums.backdropcms.org, then on client sites that had been upgraded from CKEditor 4. 
https://github.com/backdrop-ops/forum.backdropcms.org/issues/150
I am now able to recreate this problem on a brand new dev site where CKEditor 5 is the default editor from the start. 

When creating a node with an inline image and saving, there seems to be a failure to activate CKEditor 5. 

The next time you edit the node, you will get the following message:
"Activating CKEditor 5 will reformat the content of this field. Review content carefully after activating. [Click to activate editor]"(https://sunday.ddev.site/node/24/edit#).

## Steps To Reproduce

To reproduce the behavior:
1. Open a fresh demo site. 
3. Create a new post with nothing but a title and an image in the body field loaded with CKEditor 5. Save node.
4. Go to edit node and see the following:

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/9cb9ecc8-850f-467f-9dd4-6de50f312e50)

5. Note: If you save the node and then edit it again, it's fine and there is no message. 

## Expected behavior

I expect to have CKEditor already activated when I come back to edit the node. It does work as expected if there is no image in the body field. 

## Solution (Edited)

[A few comments down](https://github.com/backdrop/backdrop-issues/issues/6591#issuecomment-2156773252) we realize that when an image is initally inserted with CKEditor 5, the 'data-file-id' is listed first, before the 'src'. After activating the editor, the 'src' is moved to be first and everything is fine. 

It seems to me that we either need to change the format of the initial image insert OR make the "checker" more flexible to accept either order.

**Comment by klonos** [comment]
As I also mentioned in the Zulip thread, I was able to confirm this bug in a freshly upgraded D7 site. Not sure if it matters, but the D7 site was using CKE v3. When checking the text formats in Backdrop, they all seemed to have fallen back to using no editor. Switching to CKE5 and trying to edit various nodes sometimes showed this warning, while other times not. After checking several items of content of various types, it didn't take long to realize that this was only happening when the text fields had inline images in them.

> Note: If you save the node and then edit it again, it's fine and there is no message.

I can also confirm that 👍🏼 ...also, there is no noticeable/obvious breakage in the text markup (but I might have missed something).

**Comment by laryn** [comment]
If you do a diff on the code in the editor before and after, what’s different?

**Comment by stpaultim** [comment]
@laryn You are so smart!

This is how the image is inserted originally. 

```
<p>
    Test
</p>
<img data-file-id="19" alt="" src="/files/inline-images/angry-birds.jpeg" width="1832" height="1132">
```

After editing the node and clicking to "Activate Editor" the code looks like this. 

```
<p>
    Test
</p>
<img src="/files/inline-images/angry-birds.jpeg" data-file-id="19" alt="" width="1832" height="1132">
```

This got me thinking and I tried adding a second image to the same node, AFTER activating the editor on the original image and solving the problem. Adding a second image causes the same problem and is solved the same way. 

So the problem is not so much "activating" the editor as it is that the original input of the image uses a format that then requires "Activation" to reformat it. 

I would think this could be fixed by changing the format of the markup when the image is originally inserted OR by changing the code that is checking for CKEditor 5 to allow images with the second format.

I will update issue title and description to be more specific.

**Comment by klonos** [comment]
> ... and it should be

Not sure about that. It sounds more like that the routine/logic that checks if anything has "changed" should be adapted, in order to be looking whether the same attributes exist in the source as in the destination (regardless of their order within the `<img>` tag), rather than trying to find an exact match as it seems to be trying to do now.

**Comment by klonos** [comment]
...upon checking what [`checkValueModified()`](https://github.com/backdrop/backdrop/blob/1.x/core/modules/ckeditor5/js/ckeditor5.js#L293) does in ckeditor5.js, it seems that (at least in theory, and according to the function description) things should work:
```js
    checkValueModified: function (beforeAttachValue, afterAttachValue) {
      // Pass the before value through elementGetHtml() to standardize
      // attribute order and self-closing tags. For example, two <img> tags with
      // src, width, and height attributes should be equal, even if one uses the
      // order height, src, width. Similarly, <hr /> and <hr> should be
      // considered the same. Passing in an out of the DOM makes these two
      // values use the same order and tag closing.
```

**Comment by klonos** [comment]
This seems interesting enough. I'd like to take a quick stab at it (not an expert in JS, but trying to learn more 🙂)

**Comment by klonos** [comment]
PR up for review: https://github.com/backdrop/backdrop/pull/4780

It basically breaks down each of the before/after strings to `childNodes`, and then runs each Node through `isEqualNode()` (which ignores the order of attributes).

To test this:
1. In a vanilla Backdrop site, create a page node.
2. Give it any title.
3. Switch the body field to "Raw HTML", and add the following text to it:
   ```
   <p>
       Test
   </p>
   <img data-file-id="19" alt="" src="/files/styles/card/public/field/image/card3-organize.png" width="100" height="100">
   ```
4. Save the content.
5. Edit the content and try switching the text format to "Basic" -> warning 👎🏼 

Repeat the same in the PR sandbox. There should be no warning.

**Comment by klonos** [comment]
...I think I can convert the above to an actual test. I'll give it a go, and will ask for help if I'm stuck.

**Comment by stpaultim** [comment]
This works for me! Good work @klonos!

**Comment by keiserjb** [comment]
Hmm, this could be why I couldn't get the [Ckeditor Inline Image Style](https://backdropcms.org/project/ckeditor_inline_image_style) module to work the first time the image is added.  If the markup is different, the regex in the filter won't find the image and make the filter work.

**Comment by quicksketch** [comment]
I made a small change to the PR to use `<div>` instead of `<temp>` tags, but the result is unchanged. I tested the sandbox and confirmed the warning is no longer unnecessary shown. I tried rearranging attributes, adjusting indentation, adding wrapping `<p>` tags and everything was handled correctly without showing the warning. Adding an unsupported HTML tag _does_ show the warning as expected. Great work and nice fix @klonos!

I removed `needs - automated tests` label, while that would be awesome, we don't have JS testing currently.
